[spring-ecommerce-order - Step-2] Submit miooo0o's mission

miooo0o opened the pull request:
Hi, Dani. I'm very, very late :(
I rushed through the development and couldn't pay attention to documenting the process, so I failed to write a proper TODO list.

<img width="1224" height="229" alt="Screenshot 2025-08-25 at 16 20 40" src="https://github.com/user-attachments/assets/4289c520-5e7a-4fab-995f-63e33b8444ec" />

### I encountered several major issues initially:
1. During development, I discovered problems with `CartItem` not properly connecting to `Option`/`Product`.
2. Tables weren't being created properly, which consumed a lot of time.

Since many problems were caused by bi-directional relationships, I came to prefer uni-directional entity relations after resolving these issues.`Payment` and `Order` were designed as uni-directional based on my experience implementing the entire flow in a team project. Since I'm handling payment-order flows in the team project, I tried to create a structure that's as concise yet practical as possible for this mission.

Regarding Lock usage, I temporarily applied it based on my concurrency control experience from the team project, but reverted it because it made testing complex. I wanted to keep the mission as simple as possible.

### 1. Changed CartItem to use Option instead of Product
```kotlin
@ManyToOne(fetch = FetchType.LAZY)
var option: Option
```

### 2. Added Stripe payment system
- Execute `CheckoutController` endpoints through `StripeProperties`/`StripeClient`
- Designed to simply receive and reflect results from `Stripe` (due to time constraints)

### 3. Changed to use `.env` file
```properties
security.jwt.token.secret-key=${TOKEN_SECRET_KEY}
stripe.secret-key=${STRIPE_SECRET_KEY}
```
I will send you this .env file via Slack

### 4. Exception management using GlobalException
```kotlin
    @ExceptionHandler(
        value = [
            ProcessingException::class,
            RuntimeException::class,
            IllegalArgumentException::class,
        ],
    )
    fun handleBadRequest(e: Exception): ResponseEntity<ErrorResponse> {
        return buildErrorResponse(HttpStatus.BAD_REQUEST, e.message, e.extractErrors())
    }
```

miooo0o committed:
init files from ecommerce

miooo0o committed:
init( build.gradle.kts, application.properties ): add dependencies

miooo0o committed:
docs( README ): update readme

miooo0o committed:
refactor( Product.kt ): transfer to entity and add annotation

miooo0o committed:
refactor( Product.kt, application.properties ): match schema table columns with entity and add a property so that Hibernate does not recreate tables

BREAKING CHANGE: tests run

miooo0o committed:
refactor( Member.kt ): transfer to entity and add annotation

- rename toMemberDto to toRegisteredMember
- now Member.role defined as Role.**.name

miooo0o committed:
chore: cleanup code

miooo0o committed:
refactor( CartItem.kt, CartItemResponse.kt ): change return type in repository add/delete methods

- to apply association Product filed to CartItem entity, reduce CartItemResponse usage
- applied change to service/controller

miooo0o committed:
test: mute tests except Product, Member controller

- Decision: keep Product, Member but redesign Cart, CartItem

miooo0o committed:
refactor( CartItem.kt, CartItemResponse.kt ):

- Decision: if re-design, we decide to use `id` only depend on level
- these classes are still in progress

miooo0o committed:
refactor( CartItem.kt, Cart.kt ): change table design and entity, implement related test `CartEntityTest` with `CartRepositoryJPA`

- "spring.jpa.hibernate.ddl-auto=create-drop" to check sql table
- class CartEntityTest
- interface CartRepositoryJPA

miooo0o committed:
chore: cleanup coe

petrabenczedh committed:
refactor( model ): change entity relationships and table design

petrabenczedh committed:
docs( README ): update feature list

miooo0o committed:
chore: cleanup code

miooo0o committed:
refactor/feat( ProductRepository.kt ): use interfaces to refactor JDBC repo to JPA repo

- replace JDBC with JPA inside the ProductControllerTest.kt
- refactor functions inside the service and repository for Product

miooo0o committed:
chore: mute test

miooo0o committed:
refactor/feat( MemberRepository.kt, TokenLoginControllerTest.kt ): use interfaces to refactor JDBC repo to JPA repo

- replace JDBC with JPA inside the TokenLoginControllerTest.kt
- refactor functions inside the service and repository for Member

miooo0o committed:
chore: cleanup code

miooo0o committed:
chore: cleanup code

miooo0o committed:
Merge remote-tracking branch 'origin/step1' into step1

miooo0o committed:
refactor( CartItem ): change type of quantity to Int from Long

miooo0o committed:
chore: cleanup code

- remove unused JDBC-related files and lines
- adjust schema.sql conditions (not used with JPA)
- cleanup application.properties

miooo0o committed:
test: add a Fixture object to hold pre-made test instances

miooo0o committed:
test(service): add a new test to Cart Service

miooo0o committed:
refactor(cart): add controller with minimal service and repository for Cart/CartItem

- still need to implement

miooo0o committed:
refactor( Cart.kt ): add business logic to addItem and removeItem methods

miooo0o committed:
feat( DatabaseConfig.kt, DatabaseFixture.kt ): predefine table values for the application and for testing

- what previously has been data.sql's job is now done here
- contains constants

miooo0o committed:
style: run ktlint formatter

miooo0o committed:
chore: cleanup code

miooo0o committed:
refactor(cart): extend entity methods and adjust DTO parameters for usage

miooo0o committed:
feat(cart): add add/delete endpoints to controller and service layer

miooo0o committed:
test: implement cart controller and service basic tests

miooo0o committed:
chore: remove unused DatabaseConfig.kt

miooo0o committed:
test(cart): refactor fixture setup into reusable function

- Extract fixture creation into a helper function

miooo0o committed:
test: merge exception tests to a controller test file and separate package

miooo0o committed:
chore: cleanup code

miooo0o committed:
refactor(stat): apply JPA to stat repository

petrabenczedh committed:
refactor( TokenLoginControllerTest.kt ): remove DIRTIESCONTEXT and make test work with Fixtures

miooo0o committed:
refactor( TokenLoginControllerTest.kt ): add random port instead of the defined port, so that tests can run independently

miooo0o committed:
refactor( TokenLoginControllerTest.kt ): move to other folder

miooo0o committed:
refactor( TokenLoginControllerTest.kt ): start refactoring StatisticsE2ETest.kt

miooo0o committed:
feat( ProductController, ProductService): implement pagination for the product list

miooo0o committed:
test(ProductControllerMock): add new test to `getPages` method with MockMvc

miooo0o committed:
test(CartController): add new test to `getPages` method with MockMvc

miooo0o committed:
feat( CartController, CartService): implement pagination for the Cart list as CartItemResponse

miooo0o committed:
refactor( ProductController ): rename getPages to asPages

miooo0o committed:
feat( Option.kt ); add Option Entity and Transactional annotations

- products.forEach { println(it.options) }
is used to load options lazily

miooo0o committed:
feat( OptionRequest/Response ): implement new DTOs to add options filed in Product and apply changed to related layers

miooo0o committed:
test(product): add new test to check OptionResponse to ProductController

- create new method to create Product with Option

miooo0o committed:
chore: comment in `spring.jpa.open-in-view=false` and remove @Disabled annotation from CartEntityTest

miooo0o committed:
chore: cleanup code

- remove unused methods/properties general

miooo0o committed:
test(product-controller): refactor to use OptionRepository for option tracking

miooo0o committed:
chore: remove TODO comments for a final version

- This commit removes leftover TODO comments to prepare for the final release.
- Remaining parts that may need fixes can be identified from this commit.

miooo0o committed:
test(stat): refactor test to apply JPA

miooo0o committed:
test(product-controller): create() test doesn't check created product id

miooo0o committed:
fix(stat-service): change sort condition check `updatedAt` in CartItem

miooo0o committed:
chore: cleanup code

miooo0o committed:
refactor(cart, cart-item): change CartItem update `createdAt` and `updatedAt` menially

miooo0o committed:
docs(readme): update the progress

miooo0o committed:
refactor(product-form): rename class to ProductForm from ProductFormDto

miooo0o committed:
refactor(product, option): replace mutable vars with immutable vals where reassignment is not needed

miooo0o committed:
refactor(cart-item, option): add quantity update logic and centralize entity constants

- Added logic to update `quantity` (currently unused, prepared for future order feature)
- Moved all entity-related constants into companion objects for better organization

miooo0o committed:
refactor(cart): split large functions into smaller ones

miooo0o committed:
refactor(option-response): rename id to optionId

miooo0o committed:
chore: move DatabaseFixture to test, rename to TestFixture

miooo0o committed:
docs(readme): add feature list based on review feedback

miooo0o committed:
refactor(auth): use the Elvis operator for null checks

miooo0o committed:
refactor(cart, cart-item): delegate quantity adjustment to `CartItem` instead of changing it externally

miooo0o committed:
refactor(product): optimize read() with fetch join and readOnly transaction in repository

- Use fetch join to eagerly load options and prevent N+1 query problem
- Annotate with @Transactional(readOnly = true) for clarity and potential optimization

miooo0o committed:
refactor(product): refactor codebase

- return UpsertStatus and ProductResponse in upsert API
- Return ResponseEntity<ProductResponse> in controller instead of Unit to include more response data
- Reorder parameters for consistency
- non-nullable

miooo0o committed:
refactor(option): using `LAZY` unless eager fetching is required

miooo0o committed:
refactor(option): remove unused X-Page header

miooo0o committed:
refactor(product): prepare service-layer handling for option presence check

miooo0o committed:
refactor(product, option): delegate option binding to ProductFactory

- Removed bidirectional relation setup from Option constructor
  (no longer calls product.options.add(this) in init block)
- Introduced ProductFactory to create Product with pre-built Option entities
- Moved responsibility of binding Option to Product into Product.addOption(s)
  to centralize relation management and maintain domain invariants
- Updated ProductService to use ProductFactory instead of direct entity creation
- Added duplicate option name validation in Product entity

miooo0o committed:
test: applied previous commit changed

- delegate option binding to ProductFactory

miooo0o committed:
docs(readme): move Review.md list to README.md, remove Review.md

miooo0o committed:
refactor(product): move duplication options check logic to Factory

- request.options.distinct().size == request.options.size

miooo0o committed:
refactor(cart): page CartItems by memberId; preload product and cart.member

- Query CartItem via repository instead of loading Cart.items
- Use @EntityGraph(product, cart.member) to avoid N+1
- Keep paging at DB level (no collection fetch join)
- Tests: stub Page via PageImpl/mock; assert content[] structure

miooo0o committed:
chore: remove unused variable

miooo0o committed:
docs(readme): update the progress

miooo0o committed:
docs(readme): update the progress

miooo0o committed:
docs(readme): update the progress

miooo0o committed:
Merge remote-tracking branch 'origin/step1' into step1

miooo0o committed:
chore: remove unused methods

miooo0o committed:
chore: update gradle

- annotation("jakarta.persistence.Entity")
- annotation("jakarta.persistence.MappedSuperclass")
- annotation("jakarta.persistence.Embeddable")

miooo0o committed:
feat(client, application, dto): apply Stripe payment flow

- Basic initialization (not verified yet)
- OrderRequest/Response is incomplete

miooo0o committed:
refactor(cart-item): restrict markAsUpdated to internal calls and add Kotlin-style member getter

miooo0o committed:
feat(order, order-item): implement Order and OrderItem

- Order and OrderItem incompleted

miooo0o committed:
refactor(product): change addOption to private method

miooo0o committed:
refactor(order, order-item): store the internal data as a Long while using BigDecimal for calculations.

- Expose both major and minor values for output.
- Decided to proceed without creating a separate Money class for now. Will extract it later if needed.

miooo0o committed:
feat(order-service, repository): initiate OrderRepository and OrderService

miooo0o committed:
chore: rename TestFixture.kt tp BasicTestFixture.kt

miooo0o committed:
test(order): add tests to validate entity rules and implement a new test fixture as a helper

miooo0o committed:
refactor(order): add more validations for quantity, unitPrice and productName

miooo0o committed:
feat(dto): implement payment request/response

miooo0o committed:
refactor(cart-item):CartItem to recognize Product as an Option

- This change caused multiple issues (likely due to SQL queries), resulting in over 23 failing tests initially.
- Currently, only the CartService integration tests are still failing, but unit tests all passed

miooo0o committed:
docs(readme): add new lists for step 2, todo

miooo0o committed:
refactor(order-item): change property unitPrice type to BigDecimal and change OrderItem-Order relationship as uni-direction

miooo0o committed:
feat(order): add factory method to create Order from Cart

- Introduced `addItemsFromCart(cart)` to map and add cart items directly
- Added static factory method `fromCart(cart, currency)` for convenient order creation
- Ensures `Order` is initialized consistently from existing `Cart`

miooo0o committed:
refactor(cart-service): extract getOrCreateCartByMemberId to remove duplication

miooo0o committed:
refactor(option): rename property `quantity` to `availableStock`

miooo0o committed:
refactor(exception-handler): when an error occurs, all responses have ErrorResponse as the body

- Add a new Exception, ProcessingException to handle errors that occur in the order or payment process

miooo0o committed:
feat(order, service): create a process to process order requests in OrderService and OrderController

- Still incomplete
- Add `getCartForOrder` to retrieve validated carts from CartService.
- Add basic business rules required for Order to the service.

miooo0o committed:
feat(payment): implement a new entity Payment

miooo0o committed:
refactor: extract order validation logic to separate validator

- Move validation methods from OrderService to OrderValidator
- Enable reuse in PaymentService for payment processing

miooo0o committed:
refactor: extract order validation logic to separate validator

- Move validation methods from OrderService to OrderValidator
- Enable reuse in PaymentService for payment processing

miooo0o committed:
test(cart-service): fixed a test that wasn't properly referencing the product.

- Before this fix, the test was attempting to run the test using the product id instead of the option id, causing problems.
- This commit is related to commit #81eb2315.
- All tests now pass.

miooo0o committed:
test(cart-service): fixed a test that wasn't properly referencing the product.

- Before this fix, the test was attempting to run the test using the product id instead of the option id, causing problems.
- This commit is related to commit #81eb2315.
- All tests now pass.

miooo0o committed:
Merge remote-tracking branch 'origin/step2-rollback' into step2-rollback

# Conflicts:
#	src/main/kotlin/ecommerce/service/validator/StockValidator.kt

miooo0o committed:
feat(stripe): implement Stripe process and add required request/response DTOs

miooo0o committed:
feat(payment): implement payment service with required repositories and DTOs

miooo0o committed:
test(order): fix currency to EUR for all orders and update related tests

miooo0o committed:
feat(order): add validation process for orders

miooo0o committed:
feat(exception): add new OrderException

miooo0o committed:
feat(checkout): implement checkout controller and refine service accordingly

miooo0o committed:
test: add cleanup.sql

miooo0o committed:
chore: cleanup code

miooo0o committed:
test(checkout-service): creates minimal tests

miooo0o committed:
chore: update application.properties, gitignore

miooo0o committed:
chore: remove wildcard, make able to build

miooo0o committed:
chore: remove unused lock from option repository

miooo0o committed:
Merge branch 'miooo0o' into step2

miooo0o committed:
chore: solve conflict

miooo0o committed:
fix(checkout): cart to be emptied after checkout

- change applied in thest

da-nyee commented on src/main/kotlin/ecommerce/Application.kt:
```diff
@@ -1,8 +1,11 @@
+import org.springframework.boot.context.properties.EnableConfigurationProperties
 import org.springframework.boot.runApplication
 
+@EnableConfigurationProperties(StripeProperties::class)
```
I think it would make more sense to annotate `StripeClient` instead 🤔

miooo0o replied:
Hello. I don't quite understand this comment. 
in my understand that `@EnableConfigurationProperties` makes properties into Spring beans and automatically fills in configuration values, but if I put Client in there, what would it fetch and fill?

da-nyee replied:
I thought it would be better to keep the configuration close to where it's used.
However, as you mentioned, it still registers as a global bean, so that wouldn't really make sense.
The current version is a more appropriate approach 👍

miooo0o replied:
I see! Thank you. 
I used this annotation without much thought, was nice to be able to look it up and think about it while looking at your question :)

da-nyee commented on src/main/kotlin/ecommerce/client/StripeProperties.kt:
```diff
@@ -0,0 +1,10 @@
+@ConfigurationProperties("stripe")
+data class StripeProperties(
+    val secretKey: String,
+    // TODO: .env setting
```
Thanks for sharing the `.env` values 👍

da-nyee commented on src/main/kotlin/ecommerce/client/StripeClient.kt:
```diff
@@ -0,0 +1,54 @@
+
+@Component
+class StripeClient(private val stripeProperties: StripeProperties) {
+    private val restClient = RestClient.create()
```
What about setting timeouts for the client?

da-nyee replied:
Have you heard of `WebClient`? If so, how is it different from this one?

miooo0o replied:
```kotlin
    private val restClient = RestClient.builder()
        .requestFactory(SimpleClientHttpRequestFactory().apply {
            setConnectTimeout(Duration.ofMillis(500))
            setReadTimeout(Duration.ofSeconds(10))
        })
        .build()
```

I tried applying it like this!
Regarding WebClient, no, I heard about it for the first time. I looked it up and found out it's asynchronous and non-blocking! 
...
WebClient was introduced in Spring 5 and RestClient in Spring 6.1. After looking into the pros and cons of both, I think RestClient would have been more suitable for my project.

da-nyee commented on src/main/kotlin/ecommerce/client/StripeClient.kt:
```diff
@@ -0,0 +1,54 @@
+        }
+    }
+
+    fun validatePaymentIntent(
```
You might consider making it `private` 👀

miooo0o replied:
Thank you! Applied!

da-nyee commented on src/main/kotlin/ecommerce/client/StripeClient.kt:
```diff
@@ -0,0 +1,54 @@
+            paymentIntent
+        } catch (e: Exception) {
+            throw StripePaymentException("Stripe error: ${e.message}")
+        }
```
What about handling exceptions in more detail, following the [docs](https://docs.stripe.com/error-handling?lang=java#payment-errors)?

miooo0o replied:
I applied! 

```kotlin
sealed class StripeException(message: String, override val cause: Throwable? = null) : RuntimeException(message) {
    class Server(info: StripeErrorInfo, cause: Throwable? = null) :
        StripeException("Stripe server error: ${info.message} (code: ${info.code})", cause)

    class Client(info: StripeErrorInfo, cause: Throwable? = null) :
        StripeException("Stripe client error: ${info.message} (code: ${info.code})", cause)

    class Other(override val message: String, override val cause: Throwable? = null) : StripeException(message, cause)
}
```

I'm not sure if this turned out well. Actually, I think it would have been better if I had structured it a bit more carefully.

da-nyee replied:
You refactored the structure well.
However, what I intended was to handle exceptions such as `CardException`, `InvalidRequestException`, or others mentioned in the docs I shared.

miooo0o replied:
Oh, I see! If I had known this earlier, it could have helped my team project! Good to learn this now.

```kotlin
        } catch (e: StripeException) {
            val info = parseStripeError(e)
            when (e) {
                is CardException,
                is PermissionException,
                is RateLimitException,
                is InvalidRequestException,
                is AuthenticationException,
                is IdempotencyException,
                is SignatureVerificationException,
                -> throw StripeApiException.Client(info, e)

                is ApiConnectionException,
                is ApiException,
                -> throw StripeApiException.Server(info, e)

                else -> throw StripeApiException.Other(info, e)
            }
        } catch (e: Exception) {
            throw StripeApiException.Other(parseStripeError(e), e)
        }
```

I applied this!

da-nyee commented on src/main/kotlin/ecommerce/service/CheckoutService.kt:
```diff
@@ -0,0 +1,58 @@
+        request: CheckoutRequest,
+    ): CheckoutResponse {
+        val order = processCheckout(member, request)
+        val payment = executePayment(request)
```
Is it okay to have an external API call within a transaction? If not, what potential problems could occur?

miooo0o replied:
Hello! Thank you for the great question.

If an external API is called within a transaction, that part is outside of my control, so even if other parts are rolled back, the external API's data won't be rolled back. I didn't think about this aspect. My team project probably has the same issue (haha).

I think the solution would be to break down the current simplified process into more detailed steps, handling transactions in smaller units so that parts using external APIs are processed outside of transactions, and using event-driven asynchronous processing... but the problem would still persist. 

So combining everything, breaking transactions into smaller units while processing asynchronously to improve response speed and redesigning retry/compensation mechanisms! would be the one of approaches? 

I'm curious about your opinion...

da-nyee replied:
Yes, I think you understood it very well 👍 

That could cause data inconsistency problems, as you mentioned.
It would also lead to long response times, which may result in a poor user experience or even timeout issues.
So, it's better to separate business logic from external API call logic.

There are good approaches for this, as you commented — such as separating them into different classes or adopting an event-driven structure.

miooo0o replied:
Thank you for sharing! 😆 There are often things I wish I had known a bit earlier, and this is definitely one of them. But I'm still happy to learn about it and expand my existing knowledge! If you have any related resources or references you'd recommend, please share them...!

da-nyee commented on README.md:
```diff
@@ -1,253 +1,21 @@
-- [x] test Option
\ No newline at end of file
+### Step 2
+#### todo
```
Could you add a shell script for deployment? 👀

miooo0o replied:
I updated my readme! Thank you 👍

da-nyee reviewed (changes requested):
Hi Mina! 🙌

It's totally fine to request a late review, and you did a great job 👍 
But there are things to improve, so I left a few comments — please take a look once you have a moment.

If anything's unclear, feel free to let me know 🙇‍♀️

miooo0o committed:
refactor(stripe-client): add time out option in RestClient

miooo0o committed:
refactor(stripe-client): handle exceptions with more detail

- StripeErrorInfo parses response body from Stripe and saves it
- StripeException has more options for each exception type

miooo0o committed:
docs: update readme

miooo0o committed:
refactor(exception): global exception handle StripException

miooo0o commented:
Hi, Dani. Thank you for understanding the late submission.

I tried to apply the review you mentioned as much as possible. However, I couldn't understand the first comment well, so I wasn't able to apply it (to be more precise, it's closer to saying I tried but failed!)

I really want to hear your answer about [one of the comments](https://github.com/herotechcourse/spring-ecommerce-order/pull/34#discussion_r2301380614) you gave this time. it's a part where I could have done better in the team project... Since the project is already wrapping up, I probably won't be able to change it, but I'm still really curious.

Thank you once again for understanding the late submission.

da-nyee commented on README.md:
```diff
@@ -1,253 +1,68 @@
 
-## Feature List
-### Step 1-2 - first Refactoring
+#### start.sh
```
👍

da-nyee commented on src/main/kotlin/ecommerce/client/StripeClient.kt:
```diff
@@ -0,0 +1,90 @@
+            .requestFactory(
+                SimpleClientHttpRequestFactory().apply {
+                    setConnectTimeout(Duration.ofMillis(500))
+                    setReadTimeout(Duration.ofSeconds(10))
```
Just curious — is there any particular reason you set the timeouts to 500ms and 10s respectively?

miooo0o replied:
This isn't exactly 'the reason,' but I thought the timeout duration should be very short (especially since this checkout process wraps the entire flow and users can't interact with anything). 

Looking through the docs, it seems there are separate recommended timeouts. I fixed it by referring to this [stripe issues/817](https://github.com/stripe/stripe-java/issues/817)

```kotlin
setConnectTimeout(Duration.ofSeconds(30))
setReadTimeout(Duration.ofSeconds(80))
```

da-nyee commented on src/main/kotlin/ecommerce/exception/StripeException.kt:
```diff
@@ -0,0 +1,13 @@
+
+import ecommerce.dto.StripeErrorInfo
+
+sealed class StripeException(message: String, override val cause: Throwable? = null) : RuntimeException(message) {
+    class Server(info: StripeErrorInfo, cause: Throwable? = null) :
+        StripeException("Stripe server error: ${info.message} (code: ${info.code})", cause)
+
+    class Client(info: StripeErrorInfo, cause: Throwable? = null) :
+        StripeException("Stripe client error: ${info.message} (code: ${info.code})", cause)
+
+    class Other(override val message: String, override val cause: Throwable? = null) : StripeException(message, cause)
```
Nice use of `sealed class` 👍

miooo0o replied:
Thank you :)

miooo0o replied:
I updated this to apply your review [here](https://github.com/herotechcourse/spring-ecommerce-order/pull/34#discussion_r2304159078)

da-nyee reviewed (changes requested):
Hi Mina! 🙌

Thanks for applying the feedback 🙇‍♀️
I left a few more comments and replied to your comments, so please check them out when you have time.

It seems there's only one more step to go — keep it up until then 💪

miooo0o committed:
refactor(stripe-client): change exception logic to handle by StripeException from HttpErrorException

- add new dependency to support "StripeException"

miooo0o committed:
refactor(stripe-client): change timeout duration

- Reference: official docs 30sec/80sec

miooo0o committed:
docs(readme): update the process

miooo0o commented:
Hello Dani!
Thank you for taking the time to review. It was a huge help. 
Especially with the asynchronous, transaction, and event-driven! I was able to directly experience why such design is necessary.

### Changes Made
1. Added Dependency
```kotlin
dependencies {
    implementation("com.stripe:stripe-java:24.16.0")
    ...
```
Used this to import Stripe Exceptions.

2. Applied StripeException
```kotlin
   } catch (e: StripeException) {
            val info = parseStripeError(e)
            when (e) {
                is CardException,
                is PermissionException,
...
```
Properly parses errors from Stripe and passes them as messages.

3. Changed Stripe Client Timeout Criteria
Since it wasn't clearly documented in Stripe's documentation, I looked into it and applied it. While it's a default value and might not need to be exposed, I thought it aligned with the direction of the modifications, so I left it in.

### Brief Reflection
I really came to think that I should read documentation more carefully. Spring made me already overwhelming, and since this was my first time working with external APIs, I feel like I missed important parts -- Stripe itself. But I was genuinely happy to have learned this time.

Thanks to the feedback, I learned a lot :)

da-nyee commented on src/main/kotlin/ecommerce/client/StripeClient.kt:
```diff
@@ -0,0 +1,133 @@
+            validatePaymentIntent(request, paymentIntent)
+
+            paymentIntent
+        } catch (e: StripeException) {
+            val info = parseStripeError(e)
+            when (e) {
+                is CardException,
+                is PermissionException,
+                is RateLimitException,
+                is InvalidRequestException,
+                is AuthenticationException,
+                is IdempotencyException,
+                is SignatureVerificationException,
+                -> throw StripeApiException.Client(info, e)
+
+                is ApiConnectionException,
+                is ApiException,
+                -> throw StripeApiException.Server(info, e)
+
+                else -> throw StripeApiException.Other(info, e)
+            }
+        } catch (e: Exception) {
+            throw StripeApiException.Other(parseStripeError(e), e)
+        }
```
👍

da-nyee reviewed (approved):
Hi Mina! 🙌

It's finally time to wrap up the last mission — you nailed it 💯 
I hope the feedback was helpful for your growth, and that you keep your passion for software development.

Congratulations on your graduation! 🎉
